encoding/ccf/decode_type.go

```diff
@@ -347,9 +347,218 @@ func (d *Decoder) decodeCapabilityType(
+//	; cbor-tag-entitlement-set-authorization-type
+//	#6.146([
+//	    kind: uint8,
+//	    entitlements: +[string]
+//	])
+//
+// entitlement-map-authorization-type =
+//
+//	; cbor-tag-entitlement-map-authorization-type
+//	#6.147(entitlement: string)
+//
+// authorization-type-value =
+//
+//	unauthorized-type-value
+//	/ entitlement-set-authorization-type-value
+//	/ entitlement-map-authorization-type-value
+//
+// unauthorized-type-value = nil
+//
+// entitlement-set-authorization-type-value =
+//
+//	; cbor-tag-entitlement-set-authorization-type-value
+//	#6.195([
+//	    kind: uint8,
+//	    entitlements: +[string]
+//	])
+//
+// entitlement-map-authorization-type-value =
+//
+//	; cbor-tag-entitlement-map-authorization-type-value
+//	#6.196(entitlement: string)
+func (d *Decoder) decodeAuthorization(isType bool) (cadence.Authorization, error) {
+	nt, err := d.dec.NextType()
+	if err != nil {
+		return nil, err
+	}
+
+	switch nt {
+	case cbor.NilType:
+		err = d.dec.DecodeNil()
+		if err != nil {
+			return nil, err
+		}
+		return cadence.UnauthorizedAccess, nil
+
+	case cbor.TagType:
+		tagNum, err := d.dec.DecodeTagNumber()
+		if err != nil {
+			return nil, err
+		}
+
+		if isType {
+
+			switch tagNum {
+			case CBORTagEntitlementSetAuthorizationAccessType:
+				return d.decodeEntitlementSetAuthorization()
+
+			case CBORTagEntitlementMapAuthorizationAccessType:
+				return d.decodeEntitlementMapAuthorization()
+
+			default:
+				return nil, fmt.Errorf("unexpected CBOR tag number %d as Authorization type", tagNum)
+			}
+
+		} else {
+
+			switch tagNum {
+			case CBORTagEntitlementSetAuthorizationAccessTypeValue:
+				return d.decodeEntitlementSetAuthorization()
+
+			case CBORTagEntitlementMapAuthorizationAccessTypeValue:
+				return d.decodeEntitlementMapAuthorization()
+
+			default:
+				return nil, fmt.Errorf("unexpected CBOR tag number %d as Authorization type value", tagNum)
+			}
+
+		}
+
+	default:
+		return nil, fmt.Errorf("unexpected CBOR type %s as Authorization type", nt.String())
+	}
+}
+
+const entitlementSetAuthorizationArraySize = 2
+
+// decodeEntitlementSetAuthorization decodes
+// - entitlement-set-authorization-type, or
+// - entitlement-set-authorization-type-value
+// without CBOR tag as
+// language=CDDL
+// entitlement-set-authorization-type =
+//
+//	; cbor-tag-entitlement-set-authorization-type
+//	#6.146([
+//	    kind: uint8,
+//	    entitlements: +[string]
+//	])
+func (d *Decoder) decodeEntitlementSetAuthorization() (cadence.Authorization, error) {
+
+	// Decode array head of known length.
+	err := decodeCBORArrayWithKnownSize(d.dec, entitlementSetAuthorizationArraySize)
+	if err != nil {
+		return nil, err
+	}
+
+	// element 0: kind
+	rawKind, err := d.dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+
+	var kind cadence.EntitlementSetKind
+
+	switch rawKind {
+	case uint64(cadence.Conjunction):
+		kind = cadence.Conjunction
+	case uint64(cadence.Disjunction):
+		kind = cadence.Disjunction
```

Comment:
We might want to define an enum in CCF itself, so accidental re-ordering in the `cadence` package does not lead to change of semantics.

Reply:
Good idea!  Done in commit [e54700f](https://github.com/onflow/cadence/pull/3139/commits/e54700f90755c5402962d601e1417a2999ccfe60).